interfaces/inc/file.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef FILE_H
+#define FILE_H
+
+#ifdef __cplusplus
+#include <cstdint>
+#else
+#include <stdbool.h>
+#include <stdint.h>
+#endif
+
+#include "umock_c/umock_c_prod.h"
+#include "azure_macro_utils/macro_utils.h"
+#include "constbuffer.h"
+#include "execution_engine.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define FILE_WRITE_ASYNC_VALUES \
+    FILE_WRITE_ASYNC_INVALID_ARGS, \
+    FILE_WRITE_ASYNC_WRITE_ERROR, \
+    FILE_WRITE_ASYNC_ERROR,\
+    FILE_WRITE_ASYNC_OK
+MU_DEFINE_ENUM(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES);
+
+#define FILE_READ_ASYNC_VALUES \
+    FILE_READ_ASYNC_INVALID_ARGS, \
+    FILE_READ_ASYNC_READ_ERROR, \
+    FILE_READ_ASYNC_ERROR,\
+    FILE_READ_ASYNC_OK
+MU_DEFINE_ENUM(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES);
+
+typedef FILE_HANDLE_DATA_TAG* FILE_HANDLE;
+typedef void(*FILE_REPORT_FAULT)(void* user_report_fault_context, const char* information);
+
+typedef void(*FILE_WRITE_CB)(void* user_context, bool is_successful);
+typedef void(*FILE_READ_CB)(void* user_context, bool is_successful, CONSTBUFFER_HANDLE content);
+
+
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);

```

Comment:
>user_ [](start = 64, length = 5)

poetry: can remove the "user" part here and everywhere

Reply:
hmm i think it might be helpful to keep it to differentiate implementation specific callbacks such as on_file_io_complete from user specified callbacks


---
In reply to: [449268404](https://github.com/Azure/azure-c-pal/pull/7#discussion_r449268404) [](ancestors = 449268404)